apps/vmq_server/src/vmq_metrics_http.erl

```diff
@@ -36,33 +38,42 @@ terminate(_Reason, _Req, _State) ->
 
 reply(Req, <<"text/plain">>) ->
     %% Prometheus output
-    Metrics = vmq_metrics:metrics(true),
+    Metrics = vmq_metrics:metrics(),
     Output = prometheus_output(Metrics, []),
     {ok, Req2} = cowboy_req:reply(200, [{<<"content-type">>, <<"text/plain">>}],
                                   Output, Req),
     Req2.
 
-
-
-prometheus_output([{counter, Metric, Val, Descr}|Metrics], Acc) ->
-    BinMetric = atom_to_binary(Metric, utf8),
-    BinVal = integer_to_binary(Val),
-    Node = atom_to_binary(node(), utf8),
-    HelpLine = [<<"# HELP ">>, BinMetric, <<" ", Descr/binary, "\n">>],
-    TypeLine = [<<"# TYPE ">>, BinMetric, <<" counter\n">>],
-    CounterLine = [BinMetric, <<"{node=\"", Node/binary, "\"} ", BinVal/binary, "\n">>],
-    prometheus_output(Metrics, [HelpLine, TypeLine, CounterLine|Acc]);
-prometheus_output([{gauge, Metric, Val, Descr}|Metrics], Acc) ->
+prometheus_output([{#metric_def{type=Type, name=Metric, description=Descr, tags=Tags}, Val}|Metrics], Acc) ->
     BinMetric = atom_to_binary(Metric, utf8),
     BinVal = integer_to_binary(Val),
     Node = atom_to_binary(node(), utf8),
     HelpLine = [<<"# HELP ">>, BinMetric, <<" ", Descr/binary, "\n">>],
-    TypeLine = [<<"# TYPE ">>, BinMetric, <<" gauge\n">>],
-    CounterLine = [BinMetric, <<"{node=\"", Node/binary, "\"} ", BinVal/binary, "\n">>],
-    prometheus_output(Metrics, [HelpLine, TypeLine, CounterLine|Acc]);
+    TypeLine = [<<"# TYPE ">>, BinMetric, type(Type)],
+    Lines = lines(BinMetric, Node, Tags, BinVal),
+    prometheus_output(Metrics, [HelpLine, TypeLine, Lines|Acc]);
 prometheus_output([], Acc) -> Acc.
 
+lines(BinMetric, Node, [], BinVal) ->
+    line(BinMetric, Node, BinVal);
+lines(BinMetric, Node, Tags, BinVal) ->
+    [line(BinMetric, Node, Tag, BinVal) || Tag <- Tags].
 
+line(BinMetric, Node, BinVal) ->
+    [BinMetric, <<"{">>, label(<<"node">>, Node), <<"} ">>, BinVal, <<"\n">>].
 
+line(BinMetric, Node, Tag, BinVal) ->
+    [BinMetric,
+     <<"{">>,
+     label(<<"node">>, Node), <<",">>,
+     label(<<"tag">>, atom_to_binary(Tag, utf8)),
```

Comment:
Maybe this is overkill, but in Prometheus you usually have multiple labels per metric while this approach would limit the labels to `node=xy` and `tag=xy`. I think for our case it would be definitive enough but it would look nicer if the name `tag` isn't fixed but rather dependent on the metric... let's say for the `mqtt_*` metrics it could be `mqtt_version=[4|5]` it could even contain more labels e.g. for specifying transport protocol `[ws|wss|tcp|tls]`... but as said it's probably overkill.

Reply:
If we start differentiating the labels then we'd have to come up with sensible differentiators (which we'd need to store together with the metrics definitions).

So, for example the metric `counter.mqtt_publish_received` would have the tags `mqtt_version=4`. This could be stored as a tag k/v pair `{mqtt_version,4}` where the metric is defined and would play well with prometheus. How would this map into the graphite exporter? Maybe sth like `mqtt.publish.received.mqtt_version_4` where the k/v pair is concatenated with `_` and a metric line is created for each such k/v pair?

Also, I'm not sure how would we specify this on the CLI: `vmq-admin metrics show --tags=k=v` doesn't work as it's not currently allowed to have more than one `=` in a flag. I guess `vmq-admin metrics show --mqtt_version=4` would be the natural choice. This would work a bit like the `vmq-admin sesion show` flags.

Then the question is how this information should be displayed on the cli if not aggregating. This could be similar to the output generated for graphite, I guess (aggregation is default, so passing no flags wouldn't display any tags, only the metrics names.

What do you think?

Reply:
I think for the proposed solution for Graphite would work.. but we could also decide to do the aggregation by default for Graphite... 

I like the idea of having `vmq-admin metrics show --mqtt_version=4 --other=this ..` I think we should default to aggregate if multiple metrics for the same name with different labels exist. The case where one explicitly wants no aggregation with something like `vmq-admin metrics show --disable-aggregation` we could print them similarly as Prometheus would do e.g. inside the 'name' column `metric_name{label1=val1, label2=val2..}` and inside the `value` column the value of the metric.
 
What do you think?